README
the mg2bot
v0.71b

(c) 2008 Michael Gregorowicz
         the mg2 organization
         (mike at mg2.org)

Please see the document "LEGAL" for information pertaining to the
distribution of this software.

the mg2bot was written primarily to do channel maintenence for the channel
#mg2 on efnet.  this is the channel that our users come to to ask questions
and hang out.  the user system is currently strongly tied to our sql based
user database.  this is why many of you will find it lacking.

the thing that separates mg2bot from the rest of the perl based irc bots is
that it's based on Net::IRC, not POE.  yes i am aware this is a bad thing, but
Net::IRC is nice and simple, (albeit poorly documented.. just like my bot..)
and works just fine on my favorite irc network (efnet).

i know what the bot is missing.  it's missing plenty.  i've focused on some
areas that some of you might find useful, and i've let areas slip that some of
you might find essential.  if you find the bot to be useful, and have one or
two things you require to use the bot, let me know.  i'll probably add them if
they're not stupid (not that that's stopped me at all).  one of the bot's
strong points is the plugin system, if you are a perl programmer please
utilize them and send me what you've done.  it would be much appreciated.

below you will find release notes.  that's about all the running documentation
i've been keeping.  if you need help setting up the bot, please visit #mg2 on
efnet.  usually there is someone there who will be able to help with your
questions.

Release Notes:

First Release: 0.31pr
The first (private) release of the mg2bot. 

Last Release: 0.41pr
* Added plugin author / help / syntax information subs to the main bot
* Created plugin "plugin" to handle the interface to the new plugin info
* Implemented the first round of account inventory (user tracking) to 
allow for future enhancements that require this information.  A number of 
new handlers were created to take care of this.
* Upgraded plugins to link the IRC nick back to the mg2 username.
* Created the plugin "last" to allow users to check when their friends 
were last seen in the channel.
* Created a debugging mechanism.  (Requires an upgrade to Net::IRC::Entry 
to allow for the dump2string method to work.  Turning debugging on 
without this will CRASH the mg2bot.)
* Cleaned up the mg2bot directory structure and configuration options a bit 
more.  (Note you must set the $mg2bot_dir configuration value for the bot 
to function correctly).

Previous Release: 0.42pr
* Added local user database functionality and library.

Previous Release: 0.42pr-r2
* patched the stupid learn loophole (at the source) - Thanks tony for
fixing that up
* Fixed local user database, added country and sponsor fields

Previous Release: 0.45pr
* ripped out all my user tracking stuff from 0.42.  We now use the
lightweight "who" rather than the bulky "whois".
* got rid of "known_users" as well as join / part timestamps.  that
just seemed excessive.  if you find a use for it i will explain how
you can write a plugin to provide the same functionality
* replaced mg2bot's weather plugin with tony c's weather plugin.
* added the priority based event plugins.  This requires a bit of 
explanation so im going to go on now.  The two new plugin contexts
are "pub_last" and "priv_last".  These plugins sit at the end of the
on_public and on_msg handlers.  They're gathered by the get_plugins
function inside d.pl.  Eventually the code for this and the contexts for
all of this will be in EVERY handler.  But for now.. they're just in the
msg and public handlers.  Anyways the plugins are returned in an array
sorted by their PRIORITY.  (the p: header field).  The HIGHER the
priority number the lower the priority.  1 is the highest priority, and
well there's no limit to the lowest priority.  The plugins should
return either undef or a true value.  if a plugin returns true, it stops
running plugins and continues on.  if a plugin returns undef, it runs
the next plugin.  VERY NEET EH? ok.. thanks.
* added googlism as the first priorty based event plugin.  right now
it's priority 1.
* removed googlism code from the bot proper.
* created the %channels hash for channels and users.
* set up a $auto_reconnect flag enabling the bot to auto reconnect
* set up localize_mg2 function to localize the mg2 user database in the
mg2bot format, eliminating the total dependence on the mg2 user database
for the bots normal activities.  It can/will still op users regardless
of mg2 going down etc.
* updated status plugin
* updated userinfo plugin
* eliminated last plugin
* added the catchall_pub and catchall_priv plugins just to be cool.
* added the last handlers to contexts 'quit', 'join', 'part', and 'nick'
for quits, joins, parts, and nick changes repsectively.
* fixed the stupid channel-polluting bug.

Previous Release: 0.51pr
* added support for plugins belonging to multiple contexts.
* added configuration file (XML::Simple required)
* added channel info, bans, ops, etc.  (made the info more robust)
* added get_channel_data function to the main bot.
* chui101 moved scramble into a plugin
* updated lots of plugins to allow for private / public contexts
* worked out more data corruption bugs.
* more, but i'm too lazy to type right now.  i think these are the big
ones.

Previous Release: 0.52pr
* fixed the scramble plugin
* fixed the nick_change handler.. my dumb ass was changing all the
channels the bot was in to reflect the change.. thus anyone who changed
their nick in any of the channels suddenly showed up with their new nick
in all the channels.... this is all fixed ;)

Previous Release: 0.53b 
* implemented start and stopable plugins.  any plugin with a true value
in their t: header line will have their plugin run with "start" or
"stop" as their first argument during startup or shutdown respectively.
This can be used going forward for dependency self-checking.  Keep in
mind you should handle at least the "start" command and return a true
value or you will get a warning printed out on the console.
* wrote a plugin "log" that does what the name implies..  it logs
channel activity excluding bot-interactive commands and the bot's own
yacking.  It logs for every channel that it's in, this could still be
buggy, in fact i'm willing to wager it is.  Please let me know if you
encounter any bugs with the log plugin.
* Altered listplugins to list the plugins and the contexts they're in in
[]'s afterwards.  This is ugly and will be changed again.  The other way
was just too much spam.
* Altered plugininfo to list all contexts that a plugin is loaded into.
a rule going forward: NO PLUGIN MAY HAVE THE SAME NAME AS ANOTHER PLUGIN
NO MATTER WHAT CLASS IT GOES IN.  That's not a hard one to live by.  I
mean how long did we get by with 8x3 filenames (YAY DOS).
* Now, all output output by a plugin needs to have [pluginname]:
prefixed by the plugin.  This isn't via privmsg but printed to the bot's
console.  Debating writing a plugprint() function or ppr() which will
take care of this task for you.
* Worked really really really really really really hard to get the
channel inventory debugged and working properly.  It should be sane 99%
of the time now.  Please let me know the bot's output if the bot crashes
using the log plugin or any other plugin that uses the %channels hash.
* this is the first beta release.  please distribute discretely.

Previous Release: 0.57b
* lots of new features!!!

Previous Release: 0.60b
* Figured there have been so many releases since 0.53, I owe you guys a
bit of documentation as to what we modified.  I'll start with the most
recent additions since thats what i remember.
* Integrated Net::IRC into the bot, as well as Digest::Perl::MD5
Got rid of DBI and most traces of *SQL stuff.  the mg2 userbase
recently switched to LDAP, so I interface with my mg2 API.  So you
will notice an "mg2" dependency check at startup.
* Implemented preliminary botlink code.  We can handshake now, as well
as transfer messages across the botnet.  A few plugins were created to
serve these purposes, need_ops, op_bots.  They implement a botlink
handler that knows how to handle a request for ops.
* Fixed all sorts of network bugs by getting rid of the SIGALRM's in
lieu of creating our own primary bot loop and checking it on a basis
determined by the tick_length.
* Implemented ping_timeout config option which disconnects the bot and
reconnects if specified number of ticks passes between pings.
* Implemented per plugin config files using cf: header and XML::Simple
* Implemented mode change coordination into the mode_handler.  Bot now
dequeues queued operation that it witnesses happen.  For instance, a bot
which queued an op of another bot due to a relationship, dequeues it
when it sees it opped in the channel.  Tweak tick_length to taste.
* So so many other things.  I can't remember anymore.  If your code
broke or something tell me and I'll figure it out.  I'll be better at
documenting the next release.

Previous Release: 0.61b

CHANGES
------------------------------------------------------------------------
r1 | (no author) | 2004-03-02 13:03:14 -0500 (Tue, 02 Mar 2004) | 1 line

New repository initialized by cvs2svn.
------------------------------------------------------------------------
r2 | corrupt | 2004-03-02 13:03:14 -0500 (Tue, 02 Mar 2004) | 2 lines

Initial revision

------------------------------------------------------------------------
r6 | corrupt | 2004-03-02 13:29:13 -0500 (Tue, 02 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r7 | corrupt | 2004-03-02 13:29:32 -0500 (Tue, 02 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r8 | tony | 2004-03-02 17:06:41 -0500 (Tue, 02 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r9 | tony | 2004-03-02 17:36:05 -0500 (Tue, 02 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r10 | tony | 2004-03-02 17:43:05 -0500 (Tue, 02 Mar 2004) | 2 lines

added chagelog

------------------------------------------------------------------------
r11 | corrupt | 2004-03-02 18:53:57 -0500 (Tue, 02 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r12 | corrupt | 2004-03-02 19:22:55 -0500 (Tue, 02 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r13 | corrupt | 2004-03-03 04:43:43 -0500 (Wed, 03 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r14 | corrupt | 2004-03-03 06:44:35 -0500 (Wed, 03 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r15 | corrupt | 2004-03-03 06:58:30 -0500 (Wed, 03 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r16 | corrupt | 2004-03-03 07:43:29 -0500 (Wed, 03 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r17 | corrupt | 2004-03-03 07:45:16 -0500 (Wed, 03 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r18 | corrupt | 2004-03-03 07:47:53 -0500 (Wed, 03 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r19 | corrupt | 2004-03-04 07:09:18 -0500 (Thu, 04 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r20 | corrupt | 2004-03-04 07:31:15 -0500 (Thu, 04 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r21 | corrupt | 2004-03-04 07:34:36 -0500 (Thu, 04 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r22 | corrupt | 2004-03-04 07:40:16 -0500 (Thu, 04 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r23 | corrupt | 2004-03-04 07:45:26 -0500 (Thu, 04 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r24 | corrupt | 2004-03-04 11:40:57 -0500 (Thu, 04 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r25 | corrupt | 2004-03-04 20:06:54 -0500 (Thu, 04 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r26 | corrupt | 2004-03-04 20:13:32 -0500 (Thu, 04 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r27 | corrupt | 2004-03-05 11:50:04 -0500 (Fri, 05 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r28 | corrupt | 2004-03-05 11:55:53 -0500 (Fri, 05 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------
r29 | corrupt | 2004-03-05 17:52:38 -0500 (Fri, 05 Mar 2004) | 2 lines

*** empty log message ***

------------------------------------------------------------------------

lib/Net/IRC/DCC.pm
#####################################################################
#                                                                   #
#   Net::IRC -- Object-oriented Perl interface to an IRC server     #
#                                                                   #
#   DCC.pm: An object for Direct Client-to-Client connections.      #
#                                                                   #
#          Copyright (c) 1997 Greg Bacon & Dennis Taylor.           #
#                       All rights reserved.                        #
#                                                                   #
#      This module is free software; you can redistribute or        #
#      modify it under the terms of Perl's Artistic License.        #
#                                                                   #
#####################################################################
# $Id: DCC.pm 13 2004-03-03 09:43:43Z corrupt $

package Net::IRC::DCC;

use strict;



# --- #perl was here! ---
#
# The comments scattered throughout this module are excerpts from a
# log saved from one particularly surreal night on #perl. Ahh, the
# trials of being young, single, and drunk...
#
# ---------------------
#           \merlyn has offered the shower to a randon guy he met in a bar.
#  fimmtiu: Shower?
#           \petey raises an eyebrow at \merlyn
#  \merlyn: but he seems like a nice trucker guy...
#   archon: you offered to shower with a random guy?


# Methods that can be shared between the various DCC classes.
package Net::IRC::DCC::Connection;

use Carp;
use Socket;  # need inet_ntoa...
use strict;

sub fixaddr {
    my ($address) = @_;

    chomp $address;     # just in case, sigh.
    if ($address =~ /^\d+$/) {
        return inet_ntoa(pack "N", $address);
    } elsif ($address =~ /^[12]?\d{1,2}\.[12]?\d{1,2}\.[12]?\d{1,2}\.[12]?\d{1,2}$/) {
        return $address;
    } elsif ($address =~ tr/a-zA-Z//) {                    # Whee! Obfuscation!
        return inet_ntoa(((gethostbyname($address))[4])[0]);
    } else {
        return;
    }
}

sub bytes_in {
    return shift->{_bin};
}

sub bytes_out {
    return shift->{_bout};
}

sub nick {
    return shift->{_nick};
}

sub socket {
    return shift->{_socket};
}

sub time {
    return time - shift->{_time};
}

sub debug {
    return shift->{_debug};
}

# Changes here 1998-04-01 by MJD
# Optional third argument `$block'.
# If true, don't break the input into lines... just process it in blocks.
sub _getline {
    my ($self, $sock, $block) = @_;
    my ($input, $line);
    my $frag = $self->{_frag};

    if (defined $sock->recv($input, 10240)) {
	$frag .= $input;
	if (length($frag) > 0) {
	    
            warn "Got ". length($frag) ." bytes from $sock\n"
                if $self->{_debug};
            
	    if ($block) {          # Block mode (GET)
		return $input;
		
	    } else {               # Line mode (CHAT)
		# We're returning \n's 'cause DCC's need 'em
		my @lines = split /\012/, $frag, -1;
		$lines[-1] .= "\012";
		$self->{_frag} = ($frag !~ /\012$/) ? pop @lines : '';
		return (@lines);
	    }
	}
	else {
	    # um, if we can read, i say we should read more than 0
	    # besides, recv isn't returning undef on closed
	    # sockets.  getting rid of this connection...
	    
            warn "recv() received 0 bytes in _getline, closing connection.\n"
                if $self->{_debug};
            
	    $self->{_parent}->handler(Net::IRC::Event->new('dcc_close',
							   $self->{_nick},
							   $self->{_socket},
							   $self->{_type}));
	    $self->{_parent}->parent->removefh($sock);
	    $self->{_socket}->close;
	    $self->{_fh}->close if $self->{_fh};
	    return;
	}
    } else {
	# Error, lets scrap this connection
	
        warn "recv() returned undef, socket error in _getline()\n"
            if $self->{_debug};
	
        $self->{_parent}->handler(Net::IRC::Event->new('dcc_close',
						       $self->{_nick},
						       $self->{_socket},
						       $self->{_type}));
	$self->{_parent}->parent->removefh($sock);
	$self->{_socket}->close;
	$self->{_fh}->close if $self->{_fh};
	return;
    }
}

sub DESTROY {
    my $self = shift;
    
    # Only do the Disconnection Dance of Death if the socket is still
    # live. Duplicate dcc_close events would be a Bad Thing.
    
    if ($self->{_socket}->opened) {
	$self->{_parent}->handler(Net::IRC::Event->new('dcc_close',
						       $self->{_nick},
						       $self->{_socket},
						       $self->{_type}));
	$self->{_socket}->close;
	close $self->{_fh} if $self->{_fh};
	$self->{_parent}->{_parent}->parent->removeconn($self);
    }
    
}

sub peer {
    return ( $_[0]->{_nick}, "DCC " . $_[0]->{_type} );
}

# -- #perl was here! --
#     orev: hehe...
# Silmaril: to, not with.
#   archon: heheh
# tmtowtdi: \merlyn will be hacked to death by a psycho
#   archon: yeah, but with is much more amusing


# Connection handling GETs
package Net::IRC::DCC::GET;

use IO::Socket;
use Carp;
use strict;

@Net::IRC::DCC::GET::ISA = qw(Net::IRC::DCC::Connection);

sub new {

    my ($class, $container, $nick, $address,
	$port, $size, $filename, $handle, $offset) = @_;
    my ($sock, $fh);

    # get the address into a dotted quad
    $address = &Net::IRC::DCC::Connection::fixaddr($address);
    return if $port < 1024 or not defined $address or $size < 1;

    $fh = defined $handle ? $handle : IO::File->new(">$filename");

    unless(defined $fh) {
        carp "Can't open $filename for writing: $!";
        $sock = new IO::Socket::INET( Proto    => "tcp",
				      PeerAddr => "$address:$port" ) and
        $sock->close();
        return;
    }

    binmode $fh;                     # I love this next line. :-)
    ref $fh eq 'GLOB' ? select((select($fh), $|++)[0]) : $fh->autoflush(1);
    
    $sock = new IO::Socket::INET( Proto    => "tcp",
				  PeerAddr => "$address:$port" );

    if (defined $sock) {
	$container->handler(Net::IRC::Event->new('dcc_open',
						 $nick,
						 $sock,
						 'get',
						 'get', $sock));
	
    } else {
        carp "Can't connect to $address: $!";
        close $fh;
        return;
    }
    
    $sock->autoflush(1);

    my $self = {
	_bin        =>  defined $offset ? $offset : 0, # bytes recieved so far
        _bout       =>  0,      # Bytes we've sent
        _connected  =>  1,
	_debug      =>  $container->debug,
        _fh         =>  $fh,    # FileHandle we will be writing to.
        _filename   =>  $filename,
	_frag       =>  '',
	_nick       =>  $nick,  # Nick of person on other end
        _parent     =>  $container,
        _size       =>  $size,  # Expected size of file
        _socket     =>  $sock,  # Socket we're reading from
        _time       =>  time, 
	_type       =>  'GET',
        };

    bless $self, $class;

    return $self;
}

# -- #perl was here! --
#  \merlyn: we were both ogling a bartender named arley
#  \merlyn: I mean carle
#  \merlyn: carly
# Silmaril: man merlyn
# Silmaril: you should have offered HER the shower.
#   \petey: all three of them?

sub parse {
    my ($self) = shift;

    my $line = $self->_getline($_[0], 'BLOCKS');

    next unless defined $line;
    unless(print {$self->{_fh}} $line) {
	carp ("Error writing to " . $self->{_filename} . ": $!");
	close $self->{_fh};
	$self->{_parent}->parent->removeconn($self);
	$self->{_parent}->handler(Net::IRC::Event->new('dcc_close',
						       $self->{_nick},
						       $self->{_socket},
						       $self->{_type}));
	$self->{_socket}->close;
	return;
    }
    
    $self->{_bin} += length($line);
    
    
    # confirm the packet we've just recieved
    unless ( $self->{_socket}->send( pack("N", $self->{_bin}) ) ) {
	carp "Error writing to DCC GET socket: $!";
	close $self->{_fh};
	$self->{_parent}->parent->removeconn($self);
	$self->{_parent}->handler(Net::IRC::Event->new('dcc_close',
						       $self->{_nick},
						       $self->{_socket},
						       $self->{_type}));
	$self->{_socket}->close;
	return;
    }
    
    $self->{_bout} += 4;

    # The file is done.
    # If we close the socket, the select loop gets screwy because
    # it won't remove its reference to the socket.
    if ( $self->{_size} and $self->{_size} <= $self->{_bin} ) {
        close $self->{_fh};
        $self->{_parent}->parent->removeconn($self);
        $self->{_parent}->handler(Net::IRC::Event->new('dcc_close',
                                                       $self->{_nick},
                                                       $self->{_socket},
                                                       $self->{_type}));
	$self->{_socket}->close;
        return;
    }
    
    $self->{_parent}->handler(Net::IRC::Event->new('dcc_update',
                                                   $self->{_nick},
                                                   $self,
                                                   $self->{_type},
                                                   $self ));
}

sub filename {
    return shift->{_filename};
}

sub size {
    return shift->{_size};
}

sub close {
    my ($self, $sock) = @_;
    $self->{_fh}->close;
    $self->{_parent}->parent->removeconn($self);
    $self->{_parent}->handler(Net::IRC::Event->new('dcc_close',
                                                  $self->{_nick},
                                                  $self->{_socket},
                                                  $self->{_type}));
    $self->{_socket}->close;
    return;
}

# -- #perl was here! --
#  \merlyn: I can't type... she created a numbner of very good drinks
#  \merlyn: She's still at work
#           \petey resists mentioning that there's "No manual entry
#           for merlyn."
# Silmaril: Haven't you ever seen swingers?
#  \merlyn: she's off tomorrow... will meet me at the bar at 9:30
# Silmaril: AWWWWwwww yeeeaAAHH.
#   archon: waka chica waka chica


# Connection handling SENDs
package Net::IRC::DCC::SEND;
@Net::IRC::DCC::SEND::ISA = qw(Net::IRC::DCC::Connection);

use IO::File;
use IO::Socket;
use Carp;
use strict;

sub new {

    my ($class, $container, $nick, $filename, $blocksize) = @_;
    my ($size, $port, $fh, $sock, $select);

    $blocksize ||= 1024;

    # Shell-safe DCC filename stuff. Trying to prank-proof this
    # module is rather difficult.
    $filename =~ tr/a-zA-Z.+0-9=&()[]%\-\\\/:,/_/c;
    $fh = new IO::File $filename;

    unless (defined $fh) {
        carp "Couldn't open $filename for reading: $!";
	return;
    }

    binmode $fh;
    $fh->seek(0, SEEK_END);
    $size = $fh->tell;
    $fh->seek(0, SEEK_SET);

    $sock = new IO::Socket::INET( Proto     => "tcp",
                                  Listen    => 1);

    unless (defined $sock) {
        carp "Couldn't open DCC SEND socket: $!";
        $fh->close;
        return;
    }    

    $container->ctcp('DCC SEND', $nick, $filename, 
                     unpack("N",inet_aton($container->hostname())),
		     $sock->sockport(), $size);

    $sock->autoflush(1);

    my $self = {
        _bin        =>  0,         # Bytes we've recieved thus far
        _blocksize  =>  $blocksize,       
        _bout       =>  0,         # Bytes we've sent
	_debug      =>  $container->debug,
        _fh         =>  $fh,       # FileHandle we will be reading from.
        _filename   =>  $filename,
	_frag       =>  '',
	_nick       =>  $nick,
        _parent     =>  $container,
        _size       =>  $size,     # Size of file
        _socket     =>  $sock,     # Socket we're writing to
        _time       =>  0,         # This gets set by Accept->parse() 
	_type       =>  'SEND',
    };

    bless $self, $class;
    
    $sock = Net::IRC::DCC::Accept->new($sock, $self);

    unless (defined $sock) {
        carp "Error in accept: $!";
        $fh->close;
        return;
    }

    return $self;
}

# -- #perl was here! --
#  fimmtiu: So a total stranger is using your shower?
#  \merlyn: yes... a total stranger is using my hotel shower
#           Stupid coulda sworn \merlyn was married...
#   \petey: and you have a date.
#  fimmtiu: merlyn isn't married.
#   \petey: not a bad combo......
#  \merlyn: perhaps a adate
#  \merlyn: not maerried
#  \merlyn: not even sober. --)

sub parse {
    my ($self, $sock) = @_;
    my $size = ($self->_getline($sock, 1))[0];
    my $buf;

    # i don't know how useful this is, but let's stay consistent
    $self->{_bin} += 4;

    unless (defined $size) {
	# Dang! The other end unexpectedly canceled.
        carp (($self->peer)[1] . " connection to " .
	      ($self->peer)[0] . " lost");
	$self->{_fh}->close;
	$self->{_parent}->parent->removefh($sock);
        $self->{_parent}->handler(Net::IRC::Event->new('dcc_close',
						       $self->{_nick},
						       $self->{_socket},
						       $self->{_type}));
	$self->{_socket}->close;
	return;
    }
    
    $size = unpack("N", $size);
    
    if ($size >= $self->{_size}) {

	if ($self->{_debug}) {
	    warn "Other end acknowledged entire file ($size >= ",
		$self->{_size}, ")";
	}
        # they've acknowledged the whole file,  we outtie
        $self->{_fh}->close;
        $self->{_parent}->parent->removeconn($self);
        $self->{_parent}->handler(Net::IRC::Event->new('dcc_close',
						       $self->{_nick},
						       $self->{_socket},
						       $self->{_type}));
	$self->{_socket}->close;
        return;
    } 

    # we're still waiting for acknowledgement, 
    # better not send any more
    return if $size < $self->{_bout};

    unless (defined $self->{_fh}->read($buf,$self->{_blocksize})) {

	if ($self->{_debug}) {
	    warn "Failed to read from source file in DCC SEND!";
	}
	$self->{_fh}->close;
        $self->{_parent}->parent->removeconn($self);
	$self->{_parent}->handler(Net::IRC::Event->new('dcc_close',
						       $self->{_nick},
						       $self->{_socket},
						       $self->{_type}));
	$self->{_socket}->close;
        return;
    }

    unless($self->{_socket}->send($buf)) {

	if ($self->{_debug}) {
	    warn "send() failed horribly in DCC SEND"
	}
        $self->{_fh}->close;
        $self->{_parent}->parent->removeconn($self);
        $self->{_parent}->handler(Net::IRC::Event->new('dcc_close',
						       $self->{_nick},
						       $self->{_socket},
						       $self->{_type}));
	$self->{_socket}->close;
	return;
    }

    $self->{_bout} += length($buf);

    $self->{_parent}->handler(Net::IRC::Event->new('dcc_update',
						   $self->{_nick},
						   $self,
						   $self->{_type},
						   $self ));
    
    return 1;
}

# -- #perl was here! --
#  fimmtiu: Man, merlyn, you must be drunk to type like that. :)
#  \merlyn: too many longislands.
#  \merlyn: she made them strong
#   archon: it's a plot
#  \merlyn: not even a good amoun tof coke
#   archon: she's in league with the guy in your shower
#   archon: she gets you drunk and he takes your wallet!


# handles CHAT connections
package Net::IRC::DCC::CHAT;
@Net::IRC::DCC::CHAT::ISA = qw(Net::IRC::DCC::Connection);

use IO::Socket;
use Carp;
use strict;

sub new {

    my ($class, $container, $type, $nick, $address, $port) = @_;
    my ($sock, $self);

    if ($type) {
        # we're initiating

        $sock = new IO::Socket::INET( Proto     => "tcp",
                                      Listen    => 1);
	
        unless (defined $sock) {
            carp "Couldn't open DCC CHAT socket: $!";
            return;
        }

	$sock->autoflush(1);
        $container->ctcp('DCC CHAT', $nick, 'chat',
                         unpack("N",inet_aton($container->hostname)),
						        $sock->sockport());

	$self = {
	    _bin        =>  0,      # Bytes we've recieved thus far
	    _bout       =>  0,      # Bytes we've sent
	    _connected  =>  1,
	    _debug      =>  $container->debug,
	    _frag       =>  '',
	    _nick       =>  $nick,  # Nick of the client on the other end
	    _parent     =>  $container,
	    _socket     =>  $sock,  # Socket we're reading from
	    _time       =>  0,      # This gets set by Accept->parse()
	    _type       =>  'CHAT',
	};
	
	bless $self, $class;
	
        $sock = Net::IRC::DCC::Accept->new($sock, $self);

	unless (defined $sock) {
	    carp "Error in DCC CHAT connect: $!";
	    return;
	}

    } else {      # we're connecting

        $address = &Net::IRC::DCC::Connection::fixaddr($address);
	return if $port < 1024 or not defined $address;

        $sock = new IO::Socket::INET( Proto    => "tcp",
				      PeerAddr => "$address:$port");

        if (defined $sock) {
	    $container->handler(Net::IRC::Event->new('dcc_open',
						     $nick,
						     $sock,
						     'chat',
						     'chat', $sock));
	} else {
	    carp "Error in DCC CHAT connect: $!";
	    return;
	}

	$sock->autoflush(1);

	$self = {
	    _bin        =>  0,      # Bytes we've recieved thus far
	    _bout       =>  0,      # Bytes we've sent
	    _connected  =>  1,
	    _nick       =>  $nick,  # Nick of the client on the other end
	    _parent     =>  $container,
	    _socket     =>  $sock,  # Socket we're reading from
	    _time       =>  time,
	    _type       =>  'CHAT',
	};
	
	bless $self, $class;
	
	$self->{_parent}->parent->addfh($self->socket,
					$self->can('parse'), 'r', $self);
    }

    return $self;
}

# -- #perl was here! --
#  \merlyn: tahtd be coole
#           KTurner bought the camel today, so somebody can afford one
#           more drink... ;)
# tmtowtdi: I've heard of things like this...
#  \merlyn: as an experience. that is.
#   archon: i can think of cooler things (;
#  \merlyn: I don't realiy have that mch in my wallet.

sub parse {
    my ($self, $sock) = @_;

    foreach my $line ($self->_getline($sock)) {
	return unless defined $line;
    
	$self->{_bin} += length($line);
	
	return undef if $line eq "\012";
	$self->{_bout} += length($line);

	$self->{_parent}->handler(Net::IRC::Event->new('chat',
						       $self->{_nick},
						       $self->{_socket},
						       'chat',
						       $line));
	
	$self->{_parent}->handler(Net::IRC::Event->new('dcc_update',
						       $self->{_nick},
						       $self,
						       $self->{_type},
						       $self ));
    }
}

# Sends a message to a channel or person.
# Takes 2 args:  the target of the message (channel or nick)
#                the text of the message to send
sub privmsg {
    my ($self) = shift;

    unless (@_) {
	croak 'Not enough arguments to privmsg()';
    }
    
    # Don't send a CR over DCC CHAT -- it's not wanted.
    $self->socket->send(join('', @_) . "\012");
}


# -- #perl was here! --
#  \merlyn: this girl carly at the bar is aBABE
#   archon: are you sure? you don't sound like you're in a condition to
#           judge such things (;
# *** Stupid has set the topic on channel #perl to \merlyn is shit-faced
#     with a trucker in the shower.
# tmtowtdi: uh, yeah...
#  \merlyn: good topic


# Sockets waiting for accept() use this to shoehorn into the select loop.
package Net::IRC::DCC::Accept;

@Net::IRC::DCC::Accept::ISA = qw(Net::IRC::DCC::Connection);
use Carp;
use Socket;   # we use a lot of Socket functions in parse()
use strict;


sub new {
    my ($class, $sock, $parent) = @_;
    my ($self);

    $self = { _debug    =>  $parent->debug,
	      _nonblock =>  1,
	      _socket   =>  $sock,
	      _parent   =>  $parent,
	      _type     =>  'accept',
	  };
    
    bless $self, $class;

    # Tkil's gonna love this one. :-)   But what the hell... it's safe to
    # assume that the only thing initiating DCCs will be Connections, right?
    # Boy, we're not built for extensibility, I guess. Someday, I'll clean
    # all of the things like this up.
    $self->{_parent}->{_parent}->parent->addconn($self);
    return $self;
}

sub parse {
    my ($self) = shift;
    my ($sock);

    $sock = $self->{_socket}->accept;
    $self->{_parent}->{_socket} = $sock;
    $self->{_parent}->{_time} = time;

    if ($self->{_parent}->{_type} eq 'SEND') {
	# ok, to get the ball rolling, we send them the first packet.
	my $buf;
	unless (defined $self->{_parent}->{_fh}->
		read($buf, $self->{_parent}->{_blocksize})) {
	    return;
	}
	unless (defined $sock->send($buf)) {
	    $sock->close;
	    $self->{_parent}->{_fh}->close;
	    $self->{_parent}->{_parent}->parent->removefh($sock);
	    $self->{_parent}->handler(Net::IRC::Event->new('dcc_close',
							   $self->{_nick},
							   $self->{_socket},
							   $self->{_type}));
	    $self->{_socket}->close;
	    return;
	}
    }
    
    $self->{_parent}->{_parent}->parent->addconn($self->{_parent});
    $self->{_parent}->{_parent}->parent->removeconn($self);

    $self->{_parent}->{_parent}->handler(Net::IRC::Event->
					 new('dcc_open',
					     $self->{_parent}->{_nick},
					     $self->{_parent}->{_socket},
					     $self->{_parent}->{_type},
					     $self->{_parent}->{_type},
					     $self->{_parent}->{_socket})
					 );
}



1;


__END__

=head1 NAME

Net::IRC::DCC - Object-oriented interface to a single DCC connection

=head1 SYNOPSIS

Hard hat area: This section under construction.

=head1 DESCRIPTION

This documentation is a subset of the main Net::IRC documentation. If
you haven't already, please "perldoc Net::IRC" before continuing.

Net::IRC::DCC defines a few subclasses that handle DCC CHAT, GET, and SEND
requests for inter-client communication. DCC objects are created by
C<Connection-E<gt>new_{chat,get,send}()> in much the same way that
C<IRC-E<gt>newconn()> creates a new connection object.

=head1 METHOD DESCRIPTIONS

This section is under construction, but hopefully will be finally written up
by the next release. Please see the C<irctest> script and the source for
details about this module.

=head1 AUTHORS

Conceived and initially developed by Greg Bacon E<lt>gbacon@adtran.comE<gt> and
Dennis Taylor E<lt>dennis@funkplanet.comE<gt>.

Ideas and large amounts of code donated by Nat "King" Torkington E<lt>gnat@frii.comE<gt>.

Currently being hacked on, hacked up, and worked over by the members of the
Net::IRC developers mailing list. For details, see
http://www.execpc.com/~corbeau/irc/list.html .

=head1 URL

Up-to-date source and information about the Net::IRC project can be found at
http://netirc.betterbox.net/ .

=head1 SEE ALSO

=over

=item *

perl(1).

=item *

RFC 1459: The Internet Relay Chat Protocol

=item *

http://www.irchelp.org/, home of fine IRC resources.

=back

=cut
